### PR TITLE
Add support for linking HitShapes to turrets

### PR DESCRIPTION
**Pull request by reaperrr**
Allows to link a `HitShape` to the position and facing of a turret.

I need this downstream, and I assume some other (next-gen, a.k.a. isometric 2.5D) projects might find this useful as well, especially once directional armor comes into play.

### CODE REVIEW

**Comment by pchote** [comment]
Here's an idea on how we can do this without having to query `Turreted` every check: 3cc7d7568e641f73e6ebe5bf1dbdd89f76a88300

Would you be able to test and incorporate it into the PR?

**Comment by reaperrr** [comment]
> Would you be able to test and incorporate it into the PR?

Done. Tested it by giving both RA Cruiser turrets a rectangular HitShape, works as expected.

File: OpenRA.Mods.Common/HitShapes/Capsule.cs [hunk 1]
```diff
@@ -11,9 +11,12 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Capsule.cs]: These new using are no longer needed.

File: OpenRA.Mods.Common/HitShapes/Circle.cs [hunk 2]
```diff
@@ -11,9 +11,12 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Likewise here.

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@
 using System.Linq;
 using OpenRA.Graphics;
 using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: Likewise here.

File: OpenRA.Mods.Common/HitShapes/Rectangle.cs [hunk 4]
```diff
@@ -14,7 +14,9 @@
 using System.Linq;
 using OpenRA.Graphics;
 using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Rectangle.cs]: Likewise here.

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 5]
```diff
@@ -99,27 +100,23 @@ public WDist DistanceFromEdge(WVec v)
...
 
-		IEnumerable<IRenderable> IHitShape.RenderDebugOverlay(WorldRenderer wr, Actor actor)
+		IEnumerable<IRenderable> IHitShape.RenderDebugOverlay(WorldRenderer wr, WPos actorPos, WRot actorOrientation)
 		{
-			var actorPos = actor.CenterPosition;
-			var orientation = actor.Orientation + WRot.FromYaw(LocalYaw);
-
-			var vertsTop = combatOverlayVertsTop.Select(v => actorPos + v.Rotate(orientation)).ToArray();
-			var vertsBottom = combatOverlayVertsBottom.Select(v => actorPos + v.Rotate(orientation)).ToArray();
+			var vertsTop = combatOverlayVertsTop.Select(v => actorPos + v.Rotate(actorOrientation)).ToArray();
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: Missing `orientation += WRot.FromYaw(LocalYaw);` before this.

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 6]
```diff
@@ -20,6 +21,9 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Shape of actor for targeting and damage calculations.")]
 	public class HitShapeInfo : ConditionalTraitInfo, Requires<BodyOrientationInfo>
 	{
+		[Desc("Name of turret this shape is linked to. Leave empty to link shape to body.")]
+		public readonly string LinkedTurret = null;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: We call this just `Turret` in `Armament` and `With*Turret`, so can we please rename this to match?

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 7]
```diff
@@ -64,8 +68,9 @@ static object LoadShape(MiniYaml yaml)
 
 	public class HitShape : ConditionalTrait<HitShapeInfo>, ITargetablePositions
 	{
-		BodyOrientation orientation;
+		protected BodyOrientation orientation;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: Does this still need to be `protected`?

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 8]
```diff
@@ -74,13 +79,19 @@ protected override void Created(Actor self)
...
 		}
 
 		bool ITargetablePositions.AlwaysEnabled { get { return Info.RequiresCondition == null; } }
 
 		IEnumerable<WPos> ITargetablePositions.TargetablePositions(Actor self)
+		{
+			return TargetablePositions(self);
+		}
+
+		protected virtual IEnumerable<WPos> TargetablePositions(Actor self)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: Does this still need to be `protected virtual`?

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 9]
```diff
@@ -74,13 +79,19 @@ protected override void Created(Actor self)
 		{
 			orientation = self.Trait<BodyOrientation>();
 			targetableCells = self.TraitOrDefault<ITargetableCells>();
+			turret = self.TraitsImplementing<Turreted>().SingleOrDefault(t => t.Name == Info.LinkedTurret);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: We use `FirstOrDefault` (or just `First`) everywhere else, so best to be consistent here.

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 10]
```diff
@@ -91,9 +102,53 @@ IEnumerable<WPos> ITargetablePositions.TargetablePositions(Actor self)
 
 			foreach (var o in Info.TargetableOffsets)
 			{
-				var offset = orientation.LocalToWorld(o.Rotate(orientation.QuantizeOrientation(self, self.Orientation)));
+				var offset = CalculateTargetableOffset(self, o);
 				yield return self.CenterPosition + offset;
 			}
 		}
+
+		protected virtual WVec CalculateTargetableOffset(Actor self, WVec offset)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: Does this still need to be `protected virtual`?

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 11]
```diff
@@ -91,9 +102,53 @@ IEnumerable<WPos> ITargetablePositions.TargetablePositions(Actor self)
...
+				// then rotate back at the end
+				var turretOrientation = turret.WorldOrientation(self) - quantizedBodyOrientation;
+				localOffset = localOffset.Rotate(turretOrientation);
+				localOffset += turret.Offset;
+			}
+
+			return orientation.LocalToWorld(localOffset.Rotate(quantizedBodyOrientation));
+		}
+
+		public WDist DistanceFromEdge(Actor self, WPos pos)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: If the other methods still need to be virtual, won't these ones also need to?

**Review by pchote — COMMENTED** [review]
A few more small things which I missed the first time around:

File: OpenRA.Mods.Common/Traits/HitShape.cs (hunk 7 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/HitShape.cs]: Nah, this and the other ones basically are copy-pasta leftovers from back when downstream was using a `TurretHitShape`. I'll remove them.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 12]
```diff
@@ -74,13 +79,19 @@ protected override void Created(Actor self)
...
 		}
 
 		bool ITargetablePositions.AlwaysEnabled { get { return Info.RequiresCondition == null; } }
 
 		IEnumerable<WPos> ITargetablePositions.TargetablePositions(Actor self)
+		{
+			return TargetablePositions(self);
+		}
+
+		IEnumerable<WPos> TargetablePositions(Actor self)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: This can be merged back into `ITargetablePositions.TargetablePositions` above now.

**Review by pchote — DISMISSED** [review]
LGTM. Just one minor style suggestion before we merge.

**Comment by reaperrr** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/741ff3811e131164c36f69fb93221272023c25c0...15daf4e204e5a5ae3430d547868a0a1e813a842a)

**Pull request merged**